Quick overhaul of the distributed tracing docs (#11412)
getsentry/sentry-docs: docs/concepts/key-terms/tracing/distributed-tracing.mdx

---
title: Distributed Tracing
sidebar_order: 10

> description: "Track software performance, measure throughput & latency, and see the impact of errors across multiple systems with distributed tracing."

Reviewer: ```suggestion
description: With distributed tracing, you can track software performance and measure throughput & latency, while seeing the impact of errors across multiple systems.
```

#### Spans

Span data includes:

- `parent_span_id`: ties the span to its parent span
- `op`: short string identifying the type or category of operation the span is measuring
- `start_timestamp`: when the span was opened
- `end_timestamp`: when the span was closed
- `description`: longer description of the span's operation
- `status`: short code indicating operation's status
- `tags`: key-value pairs holding additional data about the span
- `data`: arbitrarily-structured additional data about the span

An example use of the `op` and `description` properties together is `op: db.query` and `description: SELECT * FROM users WHERE last_active < %s`. The `status` property indicates the success or failure of the span's operation, or a response code for HTTP requests. `Tags` and `data` attach further contextual information to the span, such as `function: middleware.auth.is_authenticated` for a function call or `request: {url: ..., headers: ... , body: ...}` for an HTTP request.
To search span data see [Searchable Properties](/concepts/search/searchable-properties/spans/)

### Important Information About Tracing, Spans & Transactions

#### Trace Duration

> A trace begins when the earliest transaction starts and ends when the latest transaction ends. You can't "start" or "end" a trace, but you can initiate a transaction which triggers a trace. Completing all transactions ends a trace.

Reviewer: ```suggestion
A trace begins when the earliest transaction starts, and ends when the last transaction ends. You can't "start" or "end" a trace, but you can initiate a transaction which will triggers a trace. Completing all transactions ends a trace.
```

#### Async Transactions

> Child transactions may outlive the transactions containing their parent spans. For example, if a backend API call sets off a long-running task and then immediately returns a response, the backend transaction will finish long before the async task transaction does. Asynchronicity means the order in which transactions are sent to Sentry doesn't depend on the order they were created. However, order of receipt for transactions in the same trace is correlated with order of completion.

Reviewer: ```suggestion
Child transactions may outlive the transactions containing their parent spans. For example, if a backend API call sets off a long-running task and then immediately returns a response, the backend transaction will finish long before the async task transaction does. Asynchronicity means that the order in which transactions are sent to Sentry won't depend on the order they were created. However, order of receipt for transactions in the same trace is correlated with order of completion.
```

#### Orphan Transactions

In a fully instrumented system, each trace should contain only one transaction and one span (the transaction's root) without a parent. However, you may not have tracing enabled in every service, or an instrumented service may fail to report a transaction. This can result in gaps in your trace hierarchy, with non-originating, parentless transactions called **orphan transactions**.

#### Nested Spans

There's no set limit to how deep the nesting of spans can go, but practical limits exist: transaction payloads sent to Sentry have a maximum allowed size, and there's a balance between data granularity and usability.

#### Zero-duration Spans

> A span can have equal start and end times, recorded as taking no time. This can occur if the span is used as a marker or if the operation's duration is less than the measurement resolution.

Reviewer: ```suggestion
A span can have equal start and end times which will be recorded as taking no time. This can occur if the span is being used as a marker, for example, in the browser's [performance API](https://developer.mozilla.org/en-US/docs/Web/API/Performance/mark), or if the operation's duration is less than the measurement resolution (which will vary by service).
```

#### Clock Skew

Collecting transactions from multiple machines may result in **clock skew**, where timestamps in one transaction don't align with timestamps in another. To reduce this, use Network Time Protocol (NTP) or your cloud provider's clock synchronization services.

#### How Data is Sent

Individual spans aren't sent to Sentry; the entire transaction is sent as one unit. No span data is recorded by Sentry's servers until the transaction is closed and dispatched. Transactions are still valid and sent even if they only contain their root span.

## Data Sampling

> When you enable sampling, you choose a percentage of collected transactions to send to Sentry. For example, a sampling rate of `0.25` for an endpoint receiving 1000 requests per minute results in approximately 250 transactions being sent to Sentry each minute. Sampling helps manage the number of events sent to Sentry and balance performance and volume concerns with data accuracy.

Reviewer: ```suggestion
When you enable sampling, you choose a percentage of collected transactions to send to Sentry. For example, a sampling rate of `0.25` for an endpoint receiving 1000 requests per minute results in approximately 250 transactions being sent to Sentry each minute. Sampling helps manage the number of events sent to Sentry and balances performance and volume concerns with data accuracy.
```

It's important to note that tracing is not [profiling](https://github.com/product/explore/profiling/). A profiler measures various aspects of an application's operation, resulting in a statistical summary. While both diagnose application problems, they differ in what they measure and how data is recorded.

Learn more in this [Tracing: Frontend issues with backend solutions](https://sentry.io/resources/tracing-frontend-issues-with-backend-solutions/?original_referrer=https%3A%2F%2Fblog.sentry.io%2F) workshop.

### Why Tracing?

Applications consist of interconnected components or services. For example, a modern web application may include:

- Frontend (Single-Page Application)
- Backend (REST API)
- Task Queue
- Database Server
- Cron Job Scheduler

> Each component can be instrumented individually using a Sentry SDK to capture error data or crash reports, but this doesn't provide the full picture. Tracing ties all the data together.

Reviewer: ```suggestion
Each component can be instrumented individually using a Sentry SDK to capture error data or crash reports, but this doesn't provide the full picture. Distributed tracing ties all the data together.
```

> Tracing allows you to follow a request from the [frontend](/product/sentry-basics/integrate-frontend/) to the [backend](/product/sentry-basics/integrate-backend/) and back, pulling in data from any background tasks or notification jobs that request creates. This helps correlate Sentry error reports and gain insights into which services may negatively impact your application's performance.

Reviewer: ```suggestion
Distributed tracing allows you to follow a request from the [frontend](/product/sentry-basics/integrate-frontend/) to the [backend](/product/sentry-basics/integrate-backend/) and back, pulling in data from any background tasks or notification jobs that request creates. This helps correlate Sentry error reports and gain insights into which services may negatively impact your application's performance.
```
